src/Data/Array/Mutable/Linear.hs

```diff
@@ -100,53 +105,46 @@ fromList list (f :: Array a #-> Ur b) =
     (Prelude.length list)
     (error "invariant violation: unintialized array position")
     (\arr -> f (insert arr))
-  where
-    insert :: Array a #-> Array a
-    insert = doWrites (zip list [0..])
+ where
+  insert :: Array a #-> Array a
+  insert = doWrites (zip list [0..])
 
-    doWrites :: [(a,Int)] -> Array a #-> Array a
-    doWrites [] arr = arr
-    doWrites ((a,ix):xs) arr = doWrites xs (write arr ix a)
+  doWrites :: [(a,Int)] -> Array a #-> Array a
+  doWrites [] arr = arr
+  doWrites ((a,ix):xs) arr = doWrites xs (unsafeWrite arr ix a)
 
 -- # Mutations and Reads
 -------------------------------------------------------------------------------
 
 size :: Array a #-> (Array a, Ur Int)
-size (Array size' arr) =  (Array size' arr, Ur size')
+size (Array arr) = f (Unlifted.size arr)
+ where
+  f :: (# Array# a, Ur Int #) #-> (Array a, Ur Int)
+  f (# arr, s #) = (Array arr, s)
 
 -- | Writes a value to an index of an Array. The index should be less than the
 -- arrays size, otherwise this errors.
 write :: HasCallStack => Array a #-> Int -> a -> Array a
-write = Unsafe.toLinear writeUnsafe
-  where
-    writeUnsafe :: Array a -> Int -> a -> Array a
-    writeUnsafe arr@(Array size' _) ix val
-      | indexInRange size' ix = unsafeWrite arr ix val
-      | otherwise = error "Write index out of bounds."
+write arr i x = unsafeWrite (assertIndexInRange i arr) i x
 
 -- | Same as 'write', but does not do bounds-checking. The behaviour is undefined
 -- if an out-of-bounds index is provided.
 unsafeWrite :: Array a #-> Int -> a -> Array a
-unsafeWrite (Array size' arr) ix val =
-  case Unsafe.writeMutArr arr ix val of
-    () -> Array size' arr
+unsafeWrite arr ix val =
+  withUnlifted
+    (\arr -> (# Unlifted.write ix val arr, () #))
+    arr
+    & \(a, ()) -> a
```

Comment:
This one is a bit weird, maybe optimise it later. (there is a change that GHC can optimise it on its own, but it's a tad hard to make sure)

Reply:
Sure, I can inline the `withUnlifted` call here.